### DISCUSSION TITLE
How to find key.pem in order to connect to my server

### DISCUSSION THREAD

**Discussion by AlexKolydas** [FreeOpcUa/python-opcua#1500] [General]
Hello. I am new to this so forgive me if it's something simple. 
So i want to connect to my UPS1600 machine with OPCUA and check it's status. I am using this code:
```
async def main():
    client = opcua.Client("opc.tcp://<IP_UPS>:4840", timeout=60)
    client.set_security_string("Basic256Sha256,SignAndEncrypt,certificate_that_uaexpert_uses.der ,key.pem")
    client.application_uri = "urn:<uri>:UPS1600" 
    await client.connect()
```

I checked which certificate the UAExpert uses and i am using the same in my python script. But where can i find the key.pem file?
If i run my current code i got this:

`Received an error: MessageAbort(error:StatusCode(BadCertificateInvalid), reason:None)
Received an error: MessageAbort(error:StatusCode(BadCertificateInvalid), reason:None)
Protocol Error
Traceback (most recent call last):
 ...
  File "C:\Program Files\Python39\lib\site-packages\opcua\client\ua_client.py", line 129, in _call_callback
    raise ua.UaError(
opcua.ua.uaerrors._base.UaError: No future object found for request: 0, callbacks in list are dict_keys([1])
exception calling callback for <Future at 0x1ee61977550 state=cancelled>
Traceback (most recent call last):
...
concurrent.futures._base.TimeoutError`

UAExpert connects normally. No issues

**Comment by AndreasHeine** [comment]
currently you try to travel with a stolen passport... ^^

each opc ua application needs a own pkey and certificate which matches the client settings (application uri, ip...) so the other side of the connection can check if you are who you claim to be!
you need to create your own self signed x509v3 cert with openssl!

ref: https://github.com/FreeOpcUa/python-opcua/blob/master/examples/generate_certificate.sh

**Reply by AlexKolydas** [reply]
I downloaded also Softing OPC UA Client application and i compared both certificates with UAExpert and they are the same. This certificate is from UPS device i think and UAExpert and Softing OPC UA Client are using it

**Reply by AndreasHeine** [reply]
BadCertificateInvalid -> The OPC UA Server refuse the connection
possbile reasons:
-the cert is of wrong type
-the contents of the cert does not match with the application e.g. the uri or ip or hostname
-the cert is not in the trusted list of the servers pki

**Reply by AlexKolydas** [reply]
The third point might be correct. Because on Softing OPC UA when i first started the app, i had to put the certificate into trusted. Do you know if it's possible to do it in python somehow as well?

**Reply by AndreasHeine** [reply]
this library accepts all certs by default!

**Reply by AlexKolydas** [reply]
Ok so probably the issue is because i don't have the key.pem file (?). Is there any way to create just the key from this certificate? Is the generate_certificate.sh applicable somehow?

**Reply by AndreasHeine** [reply]
private-key.pem comes first! and with that p-key you create the cert.pem! not the other way around...

in order to make it work you need create both!
you can use the generate_certificate.sh
or
there is an very old repo of me:
https://github.com/AndreasHeine/SecurePythonOpcUaClient/tree/master/x509v3
```
Generate your own Certificate
Step 1: Change ssl.conf (subjectAltname, country, organizationName, ...)
Step 2: openssl genrsa -out key.pem 2048
Step 3: openssl req -x509 -days 365 -new -out certificate.pem -key key.pem -config ssl.conf
```

**Reply by AlexKolydas** [reply]
Could you please help me with generate_certificate.sh? I'm really sorry but not sure how to work with it.

- What settings should i change? 
- What the file will look like?
- Do i just double click on it and it will generate .der and .pem files?

Here are the settings from Softing OPC UA app

![Capture](https://user-images.githubusercontent.com/37380835/211510720-ef703c25-ac5c-49c6-bc22-b2abc68cceb8.PNG)


Is this correct?
![Capture2](https://user-images.githubusercontent.com/37380835/211512092-1f416d54-8858-4618-8414-fa634b052166.PNG)


And then i double click on it? How to generate the files? 
I'm deeply sorry for asking so many questions

**Reply by AndreasHeine** [reply]
<will_add_urn_from_uaexpert> thats not ok because it does not match your opcua client urn/uri...

for the generation you need to install openssl on your pc!
so you will have the openssl-cli available!

if you are on linux you can use the .sh file!

on windows you need to:
open command promt, cd to the location where the ssl.conf is
then
type `openssl genrsa -out key.pem 2048` hit enter -> this should create the key
then
type `openssl req -x509 -days 365 -new -out certificate.pem -key key.pem -config ssl.conf` hit enter -> this should create a certificate which is valid for 365 days

use both files in you client and setup the clients application uri accroding to the uri in the cert!

**Reply by AlexKolydas** [reply]
oh thank you so much! I created them and i am able to connect to my device !!! Thank you so much for your time :) 

My code looks like this now:

```
async def main():
    client = opcua.Client("opc.tcp://<ip>:4840", timeout=60)
    client.application_uri = "urn:<ip>:UPS1600"  # Should match in your certificate
    client.set_user("admin")
    client.set_password("admin")
    client.set_security_string("Basic256Sha256,SignAndEncrypt,files\\certificate.pem,files\\key.pem")
    client.connect()
    struct = client.get_node("ns=3;i=100191")
    before = struct.get_value()
    print(before)
    print("connected")


if __name__ == '__main__':
    asyncio.run(main())

```
